sql/core/src/main/scala/org/apache/spark/sql/execution/commands/parsers.scala

```diff
@@ -0,0 +1,420 @@
+                    case Token("TOK_TABCOLVALUES", values :: Nil) =>
+                      values match {
+                        case Token("TOK_TABCOLVALUE", vals) =>
+                          vals.map(n => unquoteString(cleanIdentifier(n.text)))
+                      }
+                  }
+                  (k, unquoteString(cleanIdentifier(value.text)))
+              }.toMap
+          }
+      }
+      AlterTableSkewedLocation(tableIdent, skewedMaps)(node.source)
+
+    case Token("TOK_ALTERTABLE_ADDPARTS", addPartsArgs) :: rest =>
+      val allowExisting = base.getClauseOption("TOK_IFNOTEXISTS", addPartsArgs)
+      val parts = if (allowExisting.isDefined) {
+        addPartsArgs.tail
+      } else {
+        addPartsArgs
+      }
+
+      val partitions: ArrayBuffer[(Map[String, Option[String]], Option[String])] =
+        new ArrayBuffer()
+      var currentPart: Map[String, Option[String]] = null
+      parts.map {
+        case t @ Token("TOK_PARTSPEC", partArgs) =>
+          if (currentPart != null) {
+            partitions += ((currentPart, None))
+          }
+          currentPart = parsePartitionSpec(t).get
+        case Token("TOK_PARTITIONLOCATION", loc :: Nil) =>
+          val location = unquoteString(loc.text)
+          if (currentPart != null) {
+            partitions += ((currentPart, Some(location)))
+            currentPart = null
+          } else {
+            // We should not reach here
+            throw new AnalysisException("Partition location must follow a partition spec.")
+          }
+      }
+
+      if (currentPart != null) {
+        partitions += ((currentPart, None))
+      }
+      AlterTableAddPartition(tableIdent, partitions, allowExisting.isDefined)(node.source)
+
+    case Token("TOK_ALTERTABLE_RENAMEPART", args) :: rest =>
+      val newPartition = parsePartitionSpec(args(0))
+      AlterTableRenamePartition(tableIdent, partition.get, newPartition.get)(node.source)
+
+    case Token("TOK_ALTERTABLE_EXCHANGEPARTITION", args) :: rest =>
+      val Seq(Some(partSpec), Some(fromTable)) =
+        base.getClauses(Seq("TOK_PARTSPEC", "TOK_TABNAME"), args)
+      val partition = parsePartitionSpec(partSpec).get
+      val fromTableIdent = base.extractTableIdent(fromTable)
+      AlterTableExchangePartition(tableIdent, fromTableIdent, partition)(node.source)
+
+    case Token("TOK_ALTERTABLE_DROPPARTS", args) :: rest =>
+      val parts = args.collect {
+        case Token("TOK_PARTSPEC", partitions) =>
+          partitions.map {
+            case Token("TOK_PARTVAL", ident :: op :: constant :: Nil) =>
+              (unquoteString(cleanIdentifier(ident.text)),
+                op.text, unquoteString(cleanIdentifier(constant.text)))
+          }
+      }
+
+      val allowExisting = base.getClauseOption("TOK_IFEXISTS", args)
+      val purge = base.getClauseOption("PURGE", args)
+
+      val replication = base.getClauseOption("TOK_REPLICATION", args).map {
```

Comment:
We could simplify the pattern match, i.e.: 

    case Token("TOK_REPLICATION", replId :: metadata) =>
      (unquoteString(cleanIdentifier(replId.text)), metadata.nonEmpty)